src/Workspaces/Remote/Core/AbstractAssetProvider.cs

```diff
@@ -110,7 +112,14 @@ async Task<ImmutableArray<DocumentInfo>> CreateDocumentInfosAsync(ChecksumsAndId
         using var _ = PooledHashSet<Checksum>.GetInstance(out var checksumSet);
         checksumSet.AddAll(checksums.Children);
 
-        var results = await this.GetAssetsAsync<T>(assetPath, checksumSet, cancellationToken).ConfigureAwait(false);
-        return results.SelectAsArray(static t => t.asset);
+        var results = ImmutableArray.CreateBuilder<T>(checksumSet.Count);
```

Comment:
>ImmutableArray.CreateBuilder [](http://example.com/codeflow?start=22&length=28)

it's a bit minor, but I believe the builder itself is an allocation. ArrayBuilder or ImmutableCollectionsMarshal might be good options here

Reply:
done.